projects/pgai/pgai/sqlalchemy/__init__.py

```diff
@@ -0,0 +1,152 @@
+            _ = self.__get__(obj, self.owner)
+        # Return the actual relationship
+        return getattr(obj, f"_{self.name}_relation")
+
+    def set_schemas_correctly(self, owner: type[DeclarativeBase]) -> None:
+        table_args_schema_name = getattr(owner, "__table_args__", {}).get("schema")
+        self.target_schema = (
+            self.target_schema
+            or table_args_schema_name
+            or owner.registry.metadata.schema
+            or "public"
+        )
+
+    def create_embedding_class(
+        self, owner: type[DeclarativeBase]
+    ) -> type[EmbeddingModel[Any]]:
+        assert self.name is not None
+        table_name = self.target_table or f"{owner.__tablename__}_{self.name}_store"
+        self.set_schemas_correctly(owner)
+        class_name = f"{to_pascal_case(self.name)}Embedding"
+        registry_instance = owner.registry
+        base: type[DeclarativeBase] = owner.__base__  # type: ignore
+
+        # Get primary key information from the fully initialized model
+        mapper = inspect(owner)
+        pk_cols = mapper.primary_key
+
+        # Create the complete class dictionary
+        class_dict: dict[str, Any] = {
+            "__tablename__": table_name,
+            "registry": registry_instance,
+            # Add all standard columns
+            "embedding_uuid": mapped_column(Text, primary_key=True),
+            "chunk": mapped_column(Text, nullable=False),
+            "embedding": mapped_column(Vector(self.dimensions), nullable=False),
+            "chunk_seq": mapped_column(Integer, nullable=False),
+        }
+
+        # Add primary key columns to the dictionary
+        for col in pk_cols:
+            class_dict[col.name] = mapped_column(col.type, nullable=False)
+
+        # Create the table args with foreign key constraint
+        table_args_dict: dict[str, Any] = dict()
+        if self.target_schema and self.target_schema != owner.registry.metadata.schema:
+            table_args_dict["schema"] = self.target_schema
+
+        # Create the composite foreign key constraint
+        fk_constraint = ForeignKeyConstraint(
+            [col.name for col in pk_cols],  # Local columns
+            [
+                f"{owner.__tablename__}.{col.name}" for col in pk_cols
+            ],  # Referenced columns
+            ondelete="CASCADE",
+        )
+
+        # Add table args to class dictionary
+        class_dict["__table_args__"] = (fk_constraint, table_args_dict)
+
+        # Create the class using type()
+        Embedding = type(class_name, (base,), class_dict)
+
+        return Embedding  # type: ignore
+
+    def __get__(
+        self, obj: DeclarativeBase | None, objtype: type[DeclarativeBase] | None = None
+    ) -> type[EmbeddingModel[Any]]:
+        if not self._initialized and objtype is not None:
+            self._embedding_class = self.create_embedding_class(objtype)
+
+            # Set up relationship if requested
+            if self.add_relationship:
```

Comment:
Other suggestion, what if we simply forward all `**kwargs, *args` to the relationship function?
SQLAlchemy relationships are [extremely configurable](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params) so I think just exposing anything the user might need while transparently adding the embedding model might be the cleanest interface.

This would basically make the `Vectorizer` helper an `embedding_relationship` function? We could even rename it as such?

Reply:
If you add the relationship, does it mean that related data is prefetched with a join?

Reply:
that's exactly what we are discussing right now. If you set lazy="select" it is only fetched as soon as it is accessed. if lazy is set to "joined" it's eagerly joined. There is also "selectin" for a separate query with an in statement, and a few other options.